homeassistant/components/fan/reproduce_state.py

```diff
@@ -56,37 +62,45 @@ async def _async_reproduce_state(
         )
         return
 
-    # Return if we are already at the right state.
-    if cur_state.state == state.state and all(
-        check_attr_equal(cur_state.attributes, state.attributes, attr)
-        for attr in ATTRIBUTES
-    ):
-        return
-
-    service_data = {ATTR_ENTITY_ID: state.entity_id}
-    service_calls = {}  # service: service_data
+    service_calls: list[tuple[str, dict[str, Any]]] = []
 
     if state.state == STATE_ON:
         # The fan should be on
         if cur_state.state != STATE_ON:
-            # Turn on the fan at first
-            service_calls[SERVICE_TURN_ON] = service_data
+            # Turn on the fan with all the speed and modes settings.
+            # The `turn_on` method will figure out in which mode to
+            # turn the fan on.
+            service_calls.append(
+                (
+                    SERVICE_TURN_ON,
+                    {
+                        attr: state.attributes.get(attr)
```

Comment:
While this may iterate over more attributes, perhaps it is more natural as it is more directly showing that we're taking a subset of the existing state attributes and applying them:

```
{ k: v for k, v in state.attributes if k in SPEED_AND_MODE_ATTRIBUTES }
```

How about renaming `SPEED_AND_MODE_ATTRIBUTES` to `TURN_ON_ATTRIBUTES`?

Reply:
```
                    {
                        attr: state.attributes[attr]
                        for attr in TURN_ON_ATTRIBUTES
                        if attr in state.attributes
                    }
```